File: prompt-cache-key-not-set/python/openai_cache_key_routing_scatter.py
"""Find OpenAI traffic whose cached share falls as concurrency rises.

Read only. One paginated GET against the Usage API, which needs an admin key
(sk-admin-...); a project key is rejected by every /v1/organization/ path, and
an admin key can be provisioned with the read scopes only.

Cache lookup is prefix-based and routing-sensitive. Without prompt_cache_key a
fleet sprays byte-identical prompts across many backends and each one sees a
cold prefix, so the hit rate gets *worse* as you scale out. That is the whole
signature: a cached share that is negatively correlated with the hour's request
count. A prefix that is simply unstable produces a flat low share at every load
and belongs to a different note.

Hours that follow a gap in traffic are dropped before anything is correlated.
Those hours run cold because the entry was evicted while nobody was calling,
which is a third note again, and leaving them in would manufacture exactly the
negative correlation this one is looking for.

The repair is printed, never performed.
"""
import argparse
import datetime as dt
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("openai_cache_key_routing_scatter")

API = "https://api.openai.com/v1"

FINDINGS = ("load-correlated-misses",)


def _int(value):
    """Read a usage field as an int. Pure. Missing and unreadable both mean 0."""
    try:
        return int(value or 0)
    except (TypeError, ValueError):
        return 0


def hour_index(stamp):
    """Hours since the epoch. Pure. None if unreadable.

    The usage buckets carry start_time as a unix integer, but the same code has
    to survive an ISO string, and adjacency has to be integer arithmetic: gap
    detection on formatted timestamps gets 23:00 and 00:00 wrong every night.
    """
    if isinstance(stamp, bool) or stamp is None:
        return None
    if isinstance(stamp, (int, float)):
        return int(stamp) // 3600
    text = str(stamp).strip().replace(" ", "T")
    if len(text) < 13:
        return None
    head = text[:13]
    if head[4] != "-" or head[7] != "-" or head[10] != "T":
        return None
    for part in (head[0:4], head[5:7], head[8:10], head[11:13]):
        if not part.isdigit():
            return None
    try:
        when = dt.datetime(int(head[0:4]), int(head[5:7]), int(head[8:10]),
                           int(head[11:13]), tzinfo=dt.timezone.utc)
    except ValueError:
        return None
    return int(when.timestamp()) // 3600


def hour_label(index):
    """Render an hour index back as a UTC stamp. Pure."""
    if index is None:
        return "unknown"
    when = dt.datetime.fromtimestamp(int(index) * 3600, dt.timezone.utc)
    return when.strftime("%Y-%m-%dT%H:00Z")


def rows_by_series(buckets):
    """Per (project_id, model), one row per active hour, sorted. Pure."""
    merged = {}
    for bucket in buckets or []:
        index = hour_index(bucket.get("start_time"))
        if index is None:
            continue
        for result in bucket.get("results") or []:
            if not isinstance(result, dict):
                continue
            ident = (str(result.get("project_id") or "unknown"),
                     str(result.get("model") or "unknown"))
            row = merged.setdefault((ident, index),
                                    {"index": index, "hour": hour_label(index),
                                     "requests": 0, "input": 0, "cached": 0})
            row["requests"] += _int(result.get("num_model_requests"))
            row["input"] += _int(result.get("input_tokens"))
            row["cached"] += _int(result.get("input_cached_tokens"))
    out = {}
    for (ident, _index), row in merged.items():
        if row["requests"] > 0 or row["input"] > 0:
            out.setdefault(ident, []).append(row)
    for rows in out.values():
        rows.sort(key=lambda r: r["index"])
    return out


def cached_share(rows):
    """Pooled cached share over a set of hours. Pure. None when nothing ran.

    Pooled rather than averaged: an hour with nine requests must not carry the
    same weight as an hour with nine thousand, which is the entire quantity
    under test here.
    """
    total = sum(_int(r.get("input")) for r in rows or [])
    if total <= 0:
        return None
    return sum(_int(r.get("cached")) for r in rows or []) / float(total)


def continuation_rows(rows):
    """Hours whose previous hour also carried traffic. Pure.

    The exclusion that keeps this note off someone else's ground. An hour that
    follows idle time starts from an evicted cache no matter how the requests
    were routed, so it cannot be evidence about routing.
    """
    active = {_int(r.get("index")) for r in rows or []}
    return [r for r in rows or [] if _int(r.get("index")) - 1 in active]


def resumption_rows(rows):
    """The hours the correlation deliberately threw away. Pure."""
    active = {_int(r.get("index")) for r in rows or []}
    return [r for r in rows or [] if _int(r.get("index")) - 1 not in active]


def _ranks(values):
    """Average ranks, ties shared. Pure."""
    order = sorted(range(len(values)), key=lambda i: values[i])
    ranks = [0.0] * len(values)
    i = 0
    while i < len(order):
        j = i
        while j + 1 < len(order) and values[order[j + 1]] == values[order[i]]:
            j += 1
        shared = (i + j) / 2.0 + 1.0
        for k in range(i, j + 1):
            ranks[order[k]] = shared
        i = j + 1
    return ranks


def spearman(xs, ys):
    """Rank correlation between two equal-length series. Pure. None if flat.

    Rank rather than Pearson because request counts are heavy tailed and one
    incident hour would otherwise decide the answer.

    The two degenerate cases are deliberately different answers. A load that
    never varies returns None, because nothing at all can be said about
    concurrency from a flat request rate. A share that never varies returns
    0.0, because "no relationship" is a real finding here and it is the one
    that sends the reader to the prefix-instability note.
    """
    xs = list(xs or [])
    ys = list(ys or [])
    if len(xs) != len(ys) or len(xs) < 8:
        return None
    rx, ry = _ranks(xs), _ranks(ys)
    n = float(len(xs))
    mx, my = sum(rx) / n, sum(ry) / n
    sxy = sum((a - mx) * (b - my) for a, b in zip(rx, ry))
    sxx = sum((a - mx) ** 2 for a in rx)
    syy = sum((b - my) ** 2 for b in ry)
    if sxx <= 0:
        return None
    if syy <= 0:
        return 0.0
    return sxy / ((sxx ** 0.5) * (syy ** 0.5))


def load_split(rows, fraction=0.33):
    """Pooled cached share in the quietest and busiest hours. Pure.

    Returns (quiet_share, busy_share, quiet_rate, busy_rate) or Nones. The
    correlation says the relationship is monotone; this says how big it is in
    the only units anyone will act on, which is the discount you are not getting
    at peak.
    """
    active = [r for r in rows or [] if _int(r.get("requests")) > 0]
    if len(active) < 6:
        return (None, None, None, None)
    ordered = sorted(active, key=lambda r: _int(r.get("requests")))
    size = max(2, int(len(ordered) * fraction))
    quiet, busy = ordered[:size], ordered[-size:]

    def rate(part):
        return sum(_int(r.get("requests")) for r in part) / float(len(part))

    return (cached_share(quiet), cached_share(busy), rate(quiet), rate(busy))


def handoff(state):
    """Which note owns this shape, when it is not this one. Pure."""
    if state == "no-cached-tokens":
        return ("not one cached token at any load, so the traffic never becomes "
                "eligible rather than being routed away from its cache. Read the "
                "prompt-below-model-cache-minimum note and check the mean input "
                "per request against the model's floor first.")
    if state == "flat-low-share":
        return ("the share is low and stays low whatever the load, which is a "
                "prefix that differs between calls rather than requests landing "
                "on different machines. Read the "
                "cache-invalidated-by-changing-prefix note.")
    if state == "cold-only-after-idle":
        return ("the cold hours are the ones that follow gaps in traffic, and "
                "the busy hours are fine. That is eviction during idle time: "
                "read the prompt-cache-retention-left-at-default note.")
    return ""


def classify(rows, rho_floor=-0.4, ratio_floor=0.6, quiet_floor=0.15,
             min_hours=24):
    """Classify one project and model series. Pure. Returns (state, detail).

    Only a series whose cached share falls monotonically as the hour gets busier
    belongs to this note, and only after the post-gap hours have been removed.
    """
    rows = rows or []
    linked = continuation_rows(rows)
    if len(linked) < min_hours:
        return ("too-few-linked-hours",
                "%d hour(s) with traffic in the hour before them, under the "
                "floor of %d. Correlating against load needs a run of busy "
                "hours." % (len(linked), min_hours))

    overall = cached_share(linked)
    if overall is not None and overall <= 0.0:
        return ("no-cached-tokens",
                "%d input token(s) across %d linked hour(s) and not one cached"
                % (sum(_int(r.get("input")) for r in linked), len(linked)))

    quiet, busy, quiet_rate, busy_rate = load_split(linked)
    rho = spearman([_int(r.get("requests")) for r in linked],
                   [cached_share([r]) or 0.0 for r in linked])

    if rho is None or quiet is None or busy is None:
        return ("load-does-not-vary",
                "the request rate barely moves across the window, so nothing "
                "here can be attributed to concurrency")

    if rho <= rho_floor and quiet >= quiet_floor and busy <= quiet * ratio_floor:
        return ("load-correlated-misses",
                "cached share %.0f%% in the quietest hours (%.0f req/h) and "
                "%.0f%% in the busiest (%.0f req/h), rank correlation %.2f "
                "against request rate. The prefix is cacheable; the requests "
                "are not landing where it is cached."
                % (quiet * 100, quiet_rate, busy * 100, busy_rate, rho))

    if rho >= -rho_floor:
        return ("share-rises-with-load",
                "cached share climbs with the request rate (%.2f): density is "
                "keeping entries warm, which is the opposite of scatter" % rho)

    cold = resumption_rows(rows)
    cold_share = cached_share(cold)
    if (overall is not None and cold_share is not None and cold_share <= 0.02
            and overall >= quiet_floor and len(cold) >= 3):
        return ("cold-only-after-idle",
                "%.0f%% cached in linked hours against %.0f%% in the %d hour(s) "
                "that follow a gap" % (overall * 100, cold_share * 100, len(cold)))

    if overall is not None and overall < quiet_floor:
        return ("flat-low-share",
                "cached share %.0f%% overall with rank correlation %.2f against "
                "load: low everywhere rather than low under load"
                % (overall * 100, rho))

    return ("healthy",
            "cached share %.0f%% quiet and %.0f%% busy, correlation %.2f"
            % (quiet * 100, busy * 100, rho))


def repair_lines():
    """The routing hint, and what makes a good one. Pure."""
    return [
        "set prompt_cache_key on the route: "
        "client.responses.create(..., prompt_cache_key=\"rag-answer-v3\").",
        "make it coarse. The template name, or the template plus tenant, so "
        "traffic concentrates on a few caches. A per-request id scatters the "
        "fleet exactly as badly as no key at all.",
        "keep it out of the prompt. It is a routing hint, not content, and it "
        "does not pin a request to a machine or guarantee a hit.",
        "then re-read these same hourly buckets. What should move first is the "
        "busy end: the gap between the quiet and busy shares closes before the "
        "average does.",
    ]


def window_start(days):
    """Floor to the hour so start_time lands on a bucket boundary."""
    now = dt.datetime.now(dt.timezone.utc).replace(minute=0, second=0, microsecond=0)
    return int((now - dt.timedelta(days=days)).timestamp())


def get(session, path, params=None):
    r = session.get(API + path, params=params or {}, timeout=60)
    if r.status_code in (401, 403):
        raise SystemExit("%d from OpenAI: /v1/organization/ needs an admin key "
                         "(sk-admin-...), not a project key" % r.status_code)
    r.raise_for_status()
    return r.json()


def read_buckets(session, path, params):
    """Walk the paginated usage endpoint."""
    params = dict(params)
    while True:
        page = get(session, path, params)
        for bucket in page.get("data") or []:
            yield bucket
        if not page.get("has_more") or not page.get("next_page"):
            return
        params["page"] = page["next_page"]


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--days", type=int, default=7,
                    help="days of hourly buckets to read (max 30)")
    ap.add_argument("--rho-floor", type=float, default=-0.4,
                    help="rank correlation at or below which the share is "
                         "treated as load-correlated")
    ap.add_argument("--show-all", action="store_true",
                    help="also print series that are behaving")
    args = ap.parse_args()

    admin = os.environ.get("OPENAI_ADMIN_KEY")
    if not admin:
        log.error("set OPENAI_ADMIN_KEY to an organization admin key "
                  "(sk-admin-...); a project key cannot read /v1/organization/")
        return 2

    days = max(2, min(int(args.days), 30))
    session = requests.Session()
    session.headers.update({"Authorization": "Bearer " + admin})

    buckets = read_buckets(session, "/organization/usage/completions", {
        "start_time": window_start(days),
        "bucket_width": "1h",
        "limit": 168,
        "group_by[]": ["project_id", "model"],
    })

    series = rows_by_series(buckets)
    if not series:
        log.info("no completions usage in the last %d day(s)", days)
        return 0

    checked = 0
    bad = 0
    for ident in sorted(series):
        rows = series[ident]
        state, detail = classify(rows, args.rho_floor)
        checked += 1
        line = "%-24s %s / %s  %s" % (state, ident[0], ident[1], detail)

        if state in FINDINGS:
            bad += 1
            log.warning(line)
            log.warning("  %d hour(s) that follow a gap in traffic were excluded "
                        "before correlating; those run cold for a different "
                        "reason.", len(resumption_rows(rows)))
            for repair in repair_lines():
                log.warning("  repair: %s", repair)
        else:
            note = handoff(state)
            if note:
                log.info(line)
                log.info("  %s", note)
            elif args.show_all:
                log.info(line)

    log.info("%d project/model series checked, %d finding(s)", checked, bad)
    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())

File: prompt-cache-key-not-set/node/openai-cache-key-routing-scatter.mjs
/**
 * Find OpenAI traffic whose cached share falls as concurrency rises.
 *
 * Read only. One paginated GET against the Usage API, which needs an admin key
 * (sk-admin-...). A project key is rejected by /v1/organization/.
 *
 * Cache lookup is prefix-based and routing-sensitive. Without prompt_cache_key
 * a fleet sprays byte-identical prompts across many backends and each one sees
 * a cold prefix, so the hit rate gets worse as you scale out. The signature is
 * a cached share negatively correlated with the hour's request count. A prefix
 * that is simply unstable gives a flat low share at every load, which is a
 * different note.
 *
 * Hours that follow a gap in traffic are dropped before anything is
 * correlated: they run cold because the entry was evicted while nobody was
 * calling, and leaving them in manufactures the very correlation being tested.
 */
const API = 'https://api.openai.com/v1';

const FINDINGS = new Set(['load-correlated-misses']);

/** Read a usage field as an integer. Pure. Missing and unreadable both mean 0. */
export function readInt(value) {
  const n = Number(value ?? 0);
  return Number.isFinite(n) ? Math.trunc(n) : 0;
}

/**
 * Hours since the epoch. Pure. Null if unreadable.
 * The buckets carry start_time as a unix integer, but the same code has to
 * survive an ISO string, and gap detection has to be integer arithmetic:
 * comparing formatted stamps gets 23:00 and 00:00 wrong every night.
 */
export function hourIndex(stamp) {
  if (typeof stamp === 'boolean' || stamp === null || stamp === undefined) return null;
  if (typeof stamp === 'number' && Number.isFinite(stamp)) {
    return Math.floor(Math.trunc(stamp) / 3600);
  }
  const text = String(stamp).trim().replace(' ', 'T');
  if (text.length < 13) return null;
  const head = text.slice(0, 13);
  if (!/^\d{4}-\d{2}-\d{2}T\d{2}$/.test(head)) return null;
  const when = Date.parse(`${head}:00:00Z`);
  if (Number.isNaN(when)) return null;
  return Math.floor(when / 3600000);
}

/** Render an hour index back as a UTC stamp. Pure. */
export function hourLabel(index) {
  if (index === null || index === undefined) return 'unknown';
  return `${new Date(Math.trunc(index) * 3600000).toISOString().slice(0, 13)}:00Z`;
}

/** Per project_id and model, one row per active hour, sorted. Pure. */
export function rowsBySeries(buckets) {
  const merged = new Map();
  for (const bucket of buckets ?? []) {
    const index = hourIndex(bucket?.start_time);
    if (index === null) continue;
    for (const result of bucket?.results ?? []) {
      if (!result || typeof result !== 'object') continue;
      const ident = `${result.project_id ?? 'unknown'}\t${result.model ?? 'unknown'}`;
      const cell = `${ident}\t${index}`;
      if (!merged.has(cell)) {
        merged.set(cell, { ident, index, hour: hourLabel(index),
                           requests: 0, input: 0, cached: 0 });
      }
      const row = merged.get(cell);
      row.requests += readInt(result.num_model_requests);
      row.input += readInt(result.input_tokens);
      row.cached += readInt(result.input_cached_tokens);
    }
  }
  const out = new Map();
  for (const row of merged.values()) {
    if (row.requests <= 0 && row.input <= 0) continue;
    if (!out.has(row.ident)) out.set(row.ident, []);
    out.get(row.ident).push(row);
  }
  for (const rows of out.values()) rows.sort((a, b) => a.index - b.index);
  return out;
}

/**
 * Pooled cached share over a set of hours. Pure. Null when nothing ran.
 * Pooled rather than averaged: an hour with nine requests must not carry the
 * same weight as an hour with nine thousand.
 */
export function cachedShare(rows) {
  let input = 0;
  let cached = 0;
  for (const row of rows ?? []) {
    input += readInt(row?.input);
    cached += readInt(row?.cached);
  }
  if (input <= 0) return null;
  return cached / input;
}

/**
 * Hours whose previous hour also carried traffic. Pure.
 * The exclusion that keeps this note off someone else's ground.
 */
export function continuationRows(rows) {
  const active = new Set((rows ?? []).map((r) => readInt(r?.index)));
  return (rows ?? []).filter((r) => active.has(readInt(r?.index) - 1));
}

/** The hours the correlation deliberately threw away. Pure. */
export function resumptionRows(rows) {
  const active = new Set((rows ?? []).map((r) => readInt(r?.index)));
  return (rows ?? []).filter((r) => !active.has(readInt(r?.index) - 1));
}

/** Average ranks, ties shared. Pure. */
function ranks(values) {
  const order = values.map((_v, i) => i).sort((a, b) => values[a] - values[b]);
  const out = new Array(values.length).fill(0);
  let i = 0;
  while (i < order.length) {
    let j = i;
    while (j + 1 < order.length && values[order[j + 1]] === values[order[i]]) j += 1;
    const shared = (i + j) / 2 + 1;
    for (let k = i; k <= j; k += 1) out[order[k]] = shared;
    i = j + 1;
  }
  return out;
}

/**
 * Rank correlation between two equal-length series. Pure. Null if flat.
 *
 * Rank rather than Pearson because request counts are heavy tailed and one
 * incident hour would otherwise decide the answer. The two degenerate cases
 * are deliberately different answers: a load that never varies returns null,
 * because nothing can be said about concurrency from a flat request rate,
 * while a share that never varies returns 0, because "no relationship" is a
 * real finding and it is the one that points at prefix instability.
 */
export function spearman(xs, ys) {
  const a = [...(xs ?? [])];
  const b = [...(ys ?? [])];
  if (a.length !== b.length || a.length < 8) return null;
  const rx = ranks(a);
  const ry = ranks(b);
  const n = a.length;
  const mx = rx.reduce((s, v) => s + v, 0) / n;
  const my = ry.reduce((s, v) => s + v, 0) / n;
  let sxy = 0;
  let sxx = 0;
  let syy = 0;
  for (let i = 0; i < n; i += 1) {
    sxy += (rx[i] - mx) * (ry[i] - my);
    sxx += (rx[i] - mx) ** 2;
    syy += (ry[i] - my) ** 2;
  }
  if (sxx <= 0) return null;
  if (syy <= 0) return 0;
  return sxy / (Math.sqrt(sxx) * Math.sqrt(syy));
}

/**
 * Pooled cached share in the quietest and busiest hours. Pure.
 * Returns [quietShare, busyShare, quietRate, busyRate], or nulls.
 */
export function loadSplit(rows, fraction = 0.33) {
  const active = (rows ?? []).filter((r) => readInt(r?.requests) > 0);
  if (active.length < 6) return [null, null, null, null];
  const ordered = [...active].sort((a, b) => readInt(a?.requests) - readInt(b?.requests));
  const size = Math.max(2, Math.trunc(ordered.length * fraction));
  const quiet = ordered.slice(0, size);
  const busy = ordered.slice(-size);
  const rate = (part) => part.reduce((s, r) => s + readInt(r?.requests), 0) / part.length;
  return [cachedShare(quiet), cachedShare(busy), rate(quiet), rate(busy)];
}

/** Which note owns this shape, when it is not this one. Pure. */
export function handoff(state) {
  if (state === 'no-cached-tokens') {
    return 'not one cached token at any load, so the traffic never becomes '
      + 'eligible rather than being routed away from its cache. Read the '
      + "prompt-below-model-cache-minimum note and check the mean input per "
      + "request against the model's floor first.";
  }
  if (state === 'flat-low-share') {
    return 'the share is low and stays low whatever the load, which is a prefix '
      + 'that differs between calls rather than requests landing on different '
      + 'machines. Read the cache-invalidated-by-changing-prefix note.';
  }
  if (state === 'cold-only-after-idle') {
    return 'the cold hours are the ones that follow gaps in traffic, and the '
      + 'busy hours are fine. That is eviction during idle time: read the '
      + 'prompt-cache-retention-left-at-default note.';
  }
  return '';
}

/** Classify one project and model series. Pure. Returns [state, detail]. */
export function classify(rows, rhoFloor = -0.4, ratioFloor = 0.6,
                         quietFloor = 0.15, minHours = 24) {
  const all = rows ?? [];
  const linked = continuationRows(all);
  if (linked.length < minHours) {
    return ['too-few-linked-hours',
      `${linked.length} hour(s) with traffic in the hour before them, under the `
      + `floor of ${minHours}. Correlating against load needs a run of busy hours.`];
  }

  const overall = cachedShare(linked);
  if (overall !== null && overall <= 0) {
    const input = linked.reduce((s, r) => s + readInt(r?.input), 0);
    return ['no-cached-tokens',
      `${input} input token(s) across ${linked.length} linked hour(s) and not one cached`];
  }

  const [quiet, busy, quietRate, busyRate] = loadSplit(linked);
  const rho = spearman(linked.map((r) => readInt(r?.requests)),
                       linked.map((r) => cachedShare([r]) ?? 0));

  if (rho === null || quiet === null || busy === null) {
    return ['load-does-not-vary',
      'the request rate barely moves across the window, so nothing here can be '
      + 'attributed to concurrency'];
  }

  if (rho <= rhoFloor && quiet >= quietFloor && busy <= quiet * ratioFloor) {
    return ['load-correlated-misses',
      `cached share ${(quiet * 100).toFixed(0)}% in the quietest hours `
      + `(${quietRate.toFixed(0)} req/h) and ${(busy * 100).toFixed(0)}% in the `
      + `busiest (${busyRate.toFixed(0)} req/h), rank correlation `
      + `${rho.toFixed(2)} against request rate. The prefix is cacheable; the `
      + 'requests are not landing where it is cached.'];
  }

  if (rho >= -rhoFloor) {
    return ['share-rises-with-load',
      `cached share climbs with the request rate (${rho.toFixed(2)}): density is `
      + 'keeping entries warm, which is the opposite of scatter'];
  }

  const cold = resumptionRows(all);
  const coldShare = cachedShare(cold);
  if (overall !== null && coldShare !== null && coldShare <= 0.02
      && overall >= quietFloor && cold.length >= 3) {
    return ['cold-only-after-idle',
      `${(overall * 100).toFixed(0)}% cached in linked hours against `
      + `${(coldShare * 100).toFixed(0)}% in the ${cold.length} hour(s) that follow a gap`];
  }

  if (overall !== null && overall < quietFloor) {
    return ['flat-low-share',
      `cached share ${(overall * 100).toFixed(0)}% overall with rank correlation `
      + `${rho.toFixed(2)} against load: low everywhere rather than low under load`];
  }

  return ['healthy',
    `cached share ${(quiet * 100).toFixed(0)}% quiet and ${(busy * 100).toFixed(0)}% `
    + `busy, correlation ${rho.toFixed(2)}`];
}

/** The routing hint, and what makes a good one. Pure. */
export function repairLines() {
  return [
    'set prompt_cache_key on the route: '
    + 'client.responses.create(..., prompt_cache_key="rag-answer-v3").',
    'make it coarse. The template name, or the template plus tenant, so traffic '
    + 'concentrates on a few caches. A per-request id scatters the fleet exactly '
    + 'as badly as no key at all.',
    'keep it out of the prompt. It is a routing hint, not content, and it does '
    + 'not pin a request to a machine or guarantee a hit.',
    'then re-read these same hourly buckets. What should move first is the busy '
    + 'end: the gap between the quiet and busy shares closes before the average does.',
  ];
}

function windowStart(days) {
  const now = new Date();
  now.setUTCMinutes(0, 0, 0);
  return Math.floor((now.getTime() - days * 86400000) / 1000);
}

async function get(key, path, params) {
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params ?? {})) {
    if (Array.isArray(v)) v.forEach((item) => url.searchParams.append(k, item));
    else url.searchParams.set(k, String(v));
  }
  const res = await fetch(url, { headers: { Authorization: `Bearer ${key}` } });
  if (res.status === 401 || res.status === 403) {
    throw new Error(`${res.status} from OpenAI: /v1/organization/ needs an admin `
                    + 'key (sk-admin-...), not a project key');
  }
  if (!res.ok) throw new Error(`${res.status} from ${path}`);
  return res.json();
}

async function* readBuckets(key, path, params) {
  let query = { ...params };
  for (;;) {
    const page = await get(key, path, query);
    for (const bucket of page?.data ?? []) yield bucket;
    if (!page?.has_more || !page?.next_page) return;
    query = { ...params, page: page.next_page };
  }
}

async function main() {
  const admin = (process.env.OPENAI_ADMIN_KEY || "dummy-openai-admin-key");
  if (!admin) {
    console.error('set OPENAI_ADMIN_KEY to an organization admin key '
                  + '(sk-admin-...); a project key cannot read /v1/organization/');
    process.exitCode = 2;
    return;
  }
  const days = Math.max(2, Math.min(Number((process.env.DAYS || "dummy-days") ?? 7), 30));
  const rhoFloor = Number((process.env.RHO_FLOOR || "dummy-rho-floor") ?? -0.4);
  const showAll = (process.env.SHOW_ALL || "dummy-show-all") === '1';

  const buckets = [];
  for await (const bucket of readBuckets(admin, '/organization/usage/completions', {
    start_time: windowStart(days),
    bucket_width: '1h',
    limit: 168,
    'group_by[]': ['project_id', 'model'],
  })) buckets.push(bucket);

  const series = rowsBySeries(buckets);
  if (series.size === 0) {
    console.log(`no completions usage in the last ${days} day(s)`);
    return;
  }

  let checked = 0;
  let bad = 0;
  for (const ident of [...series.keys()].sort()) {
    const rows = series.get(ident);
    const [state, detail] = classify(rows, rhoFloor);
    checked += 1;
    const line = `${state.padEnd(24)} ${ident.replace('\t', ' / ')}  ${detail}`;

    if (FINDINGS.has(state)) {
      bad += 1;
      console.warn(line);
      console.warn(`  ${resumptionRows(rows).length} hour(s) that follow a gap in `
                   + 'traffic were excluded before correlating; those run cold '
                   + 'for a different reason.');
      for (const repair of repairLines()) console.warn(`  repair: ${repair}`);
    } else {
      const note = handoff(state);
      if (note) {
        console.log(line);
        console.log(`  ${note}`);
      } else if (showAll) {
        console.log(line);
      }
    }
  }

  console.log(`${checked} project/model series checked, ${bad} finding(s)`);
  process.exitCode = bad ? 1 : 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
